File: src-tauri/src/domain/customer.rs
use serde::{Deserialize, Serialize};

use crate::domain::orders::OrderType;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum CustomerPhase {
    Idle,
    Order,
    Payment,
    ThankYou,
}

impl Default for CustomerPhase {
    fn default() -> Self {
        Self::Idle
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CustomerItemToppingView {
    pub name: String,
    pub price: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CustomerOrderItemView {
    pub name: String,
    pub size: Option<String>,
    pub sugar: Option<String>,
    pub ice: Option<String>,
    pub qty: i64,
    pub line_total: i64,
    pub toppings: Vec<CustomerItemToppingView>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CustomerOrderView {
    pub code: String,
    pub r#type: OrderType,
    pub table_name: Option<String>,
    pub items: Vec<CustomerOrderItemView>,
    pub subtotal: i64,
    pub discount_total: i64,
    pub total: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CustomerPaymentView {
    pub qr_svg: String,
    pub amount: i64,
    pub content: String,
    pub bank_name: String,
    pub account_number: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CustomerView {
    pub phase: CustomerPhase,
    pub order: Option<CustomerOrderView>,
    pub payment: Option<CustomerPaymentView>,
}

impl Default for CustomerView {
    fn default() -> Self {
        Self {
            phase: CustomerPhase::Idle,
            order: None,
            payment: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PaymentQr {
    pub qr_svg: String,
    pub amount: i64,
    pub content: String,
    pub bank_name: String,
    pub account_number: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Bank {
    pub bin: String,
    pub name: String,
    pub short_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Branding {
    pub shop_name: String,
    pub shop_address: String,
    pub shop_phone: String,
    pub brand_color: String,
    pub logo_path: Option<String>,
    pub idle_bg_path: Option<String>,
    pub promo_images: Vec<String>,
    pub customer_welcome_text: String,
    pub bill_footer: String,
}

impl Default for Branding {
    fn default() -> Self {
        Self {
            shop_name: "HiGi Coffee".into(),
            shop_address: String::new(),
            shop_phone: String::new(),
            brand_color: "#6F4E37".into(),
            logo_path: None,
            idle_bg_path: None,
            promo_images: Vec::new(),
            customer_welcome_text: "Chao mung quy khach".into(),
            bill_footer: "Cam on quy khach!".into(),
        }
    }
}

File: src/lib/api/customer.ts
import { invoke } from "@tauri-apps/api/core"
import type { OrderType } from "@/lib/api/orders"

export type CustomerPhase = "idle" | "order" | "payment" | "thankyou"

export type CustomerItemToppingView = {
  name: string
  price: number
}

export type CustomerOrderItemView = {
  name: string
  size: string | null
  sugar: string | null
  ice: string | null
  qty: number
  line_total: number
  toppings: CustomerItemToppingView[]
}

export type CustomerOrderView = {
  code: string
  type: OrderType
  table_name: string | null
  items: CustomerOrderItemView[]
  subtotal: number
  discount_total: number
  total: number
}

export type CustomerPaymentView = {
  qr_svg: string
  amount: number
  content: string
  bank_name: string
  account_number: string
}

export type CustomerView = {
  phase: CustomerPhase
  order: CustomerOrderView | null
  payment: CustomerPaymentView | null
}

export type PaymentQr = CustomerPaymentView

export const openCustomerDisplay = () => invoke<void>("open_customer_display")

export const closeCustomerDisplay = () => invoke<void>("close_customer_display")

export const getCustomerView = () => invoke<CustomerView>("get_customer_view")

export const setCustomerOrder = (orderId: number, phase: CustomerPhase) =>
  invoke<void>("set_customer_order", { orderId, phase })

export const setCustomerPhase = (phase: CustomerPhase) =>
  invoke<void>("set_customer_phase", { phase })

export const getPaymentQr = (orderId: number) => invoke<PaymentQr>("get_payment_qr", { orderId })
